File: problem_020.c
// -*- compile-command: "gcc -std=c99 -o problem_020 problem_020.c ceuler.c -Wall -lm" -*-
// Copyright (c) 2014 Michael Caldwell
#include <stdio.h>
#include "ceuler.h"

int main()
{
    printf("Project Euler - Problem 20:\n"
           "Find the sum of the digits in the number 100!\n\n");

    // Begin time tracking
    struct timeval start;
    gettimeofday(&start, NULL);

    int bigNum[1000] = {0};
    bool debug = false;

    bigNum[997]=1; // Starting value is 100

    for (int factorial=99 ; factorial>0 ; factorial--)
    {
        // Multiply by factorial
        for (int i=999 ; i>=0 ; i--)
            bigNum[i] *= factorial;

        // Carry-over each element if needed
        for (int i=999 ; i>=0 ; i--)
            while (bigNum[i] > 9)
            {
                bigNum[i]   -= 10;
                bigNum[i-1] += 1;
            }

        // DEBUG - Print result
        if (debug == true)
        {
            int i = 0;
            while (bigNum[++i] == 0);
            for (; i<1000 ; i++)
                printf("%d", bigNum[i]);
            printf("\n");
        }
    }

    int sum = 0;
    for (int i=0 ; i<1000 ; i++)
        sum += bigNum[i];

    printf("%d\n", sum);
    printElapsedTime(start);

    return 0;
}

File: problem_020.rs
// -*- compile-command: "rustc -o problem_020_rs problem_020.rs" -*-
// Copyright (c) 2016 Michael Caldwell

fn main() {
    println!("Project Euler - Problem 20:");
    println!("Find the sum of the digits in the number 100!\n");

    let mut bignum = Vec::new();
    bignum.push(100);

    for i in (1..100).rev() {
        for e in 0..bignum.len() {
            bignum[e] *= i;
        }

        let mut t = 0;
        while t < bignum.len() {
            while bignum[t] > 9 {
                if t == bignum.len()-1 {
                    bignum.push(0);
                }
                bignum[t] -= 10;
                bignum[t+1] += 1;
            }
            t += 1;
        }
    }

    let mut total = 0;
    for e in (0..bignum.len()).rev() {
        total += bignum[e];
        //print!("{}", bignum[e]); //DEBUG
    }

    println!("\n\n{}", total);
}
